doc/patool.1
.\"                                      -*- nroff -*-
.\" Copyright (C) 2010-2026 Bastian Kleineidam
.\"
.\" This program is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.TH PATOOL "1" "January 2026" "patool"
.SH NAME
patool - portable archive file manager for the command line
.SH SYNOPSIS
 \fBpatool\fP [\fIglobal-options\fP] (\fBlist\fP|\fBtest\fP|\fBextract\fP|\fBcreate\fP|\fBdiff\fP|\fBsearch\fP|\fBrepack\fP|\fBformats\fP|\fBversion\fP) [\fIcommand-options\fP] [\fIcommand-arguments\fP]...
.SH DESCRIPTION
Various archive formats can be created, extracted, tested, listed, searched,
repacked and compared by
\fBpatool\fP. The advantage of patool is its simplicity in handling archive
files without having to remember a myriad of programs and options.
.PP
The archive format is determined by the file(1) program and as a fallback
by the archive file extension.
.PP
\fBpatool\fP supports 7z (.7z, .cb7), ACE (.ace, .cba), ADF (.adf),
ALZIP (.alz), APE (.ape), AR (.a), ARC (.arc), ARJ (.arj), BZIP2 (.bz2),
BZIP3 (.bz3), CAB (.cab), CHM (.chm), COMPRESS (.Z), CPIO (.cpio),
DEB (.deb), DMS (.dms),
FLAC (.flac), FREEARC (.arc), GZIP (.gz), ISO (.iso), LRZIP (.lrz), LZH (.lha, .lzh),
LZIP (.lz), LZMA (.lzma), LZOP (.lzo), RPM (.rpm), RAR (.rar, .cbr),
RZIP (.rz), SHN (.shn), TAR (.tar, .cbt), UDF (.udf), XZ (.xz), ZIP (.zip, .jar, .cbz),
ZOO (.zoo), ZPAQ (.zpaq) and ZSTANDARD (.zst) archive formats.
.br
It relies on helper applications to handle those archive formats
(for example \fBxz\fP for XZ (.xz) archives).
.PP
The archive formats BZIP2, GZIP, TAR, XZ and ZIP
are supported natively and do not require helper applications to be
installed.
When using Python >= 3.14 the ZSTANDARD archive format is also supported natively.

.SH EXAMPLES
  \fBpatool extract archive.zip otherarchive.rar\fP
  \fBpatool \-\-verbose test dist.tar.gz\fP
  \fBpatool list package.deb\fP
  \fPpatool \-\-verbose create myfiles.zip file1.txt dir/\fP
  \fBpatool diff release1.0.tar.xz release2.0.zip\fP
  \fBpatool search "def urlopen" python\-3.3.tar.gz\fP
  \fBpatool repack linux\-2.6.33.tar.gz linux\-2.6.33.tar.bz2\fP
.SH GLOBAL OPTIONS
.TP
\fB\-v\fP, \fB\-\-verbose\fP
Display more info about what patool does, and display the output
of helper applications. Can be given multiple times to increase
the output even more.
.TP
\fB\-q\fP, \fB\-\-quiet\fP
Work quietly, except the commands \fBformats\fP and \fBversion\fP
still print their output. Conflicts with \fB\-\-verbose\fP.
If given twice suppresses error output from archive commands.
.TP
\fB\-\-non\-interactive\fP
Try to prevent any interactive user input (i.e. prompting for passwords
or for overwriting duplicate files). Use this option with care since
overwriting files or ignoring password prompts could lead to unintended
consequences.
.br
This only works for archive programs that allow preventing user prompts. Currently
those are \fBarj\fP, \fB7z\fP and \fBrar\fP.
.SH COMMANDS
The following rules apply to all commands:
.IP "\(bu" 4
Existing files are never overwritten.
.IP "\(bu" 4
The original archive will never be removed.
.IP "\(bu" 4
Files outside the output directory will never be created. This relies on
archive program options to prevent unpacking of files with an absolute
path name (e.g. \fB\-\-no\-absolute\-filenames\fP for \fBcpio(1)\fP).
.PP
The following commands are available.
.SS extract
\fBpatool\fP \fBextract\fP [\fB\-\-outdir\fP \fIdirectory\fP] [\fB\-\-password\fP \fIpassword\fP] <\fIarchive\fP>...
.PP
Extract files from given archives. The original archives will never
be removed and are left as is.
.TP
\fB\-\-outdir\fP \fIdirectory\fP
Extract to the given output directory. Default is to extract to
the current working directory.
.TP
\fB\-\-password\fP \fIpassword\fP
Use given password for password-protected archives.
.PP
If the archive contains exactly one
file or directory, the archive contents are extracted directly to the
output directory.
Else the files are extracted in a newly created subdirectory of the output
directory. The new directory is named after the archive filename without
the extension.
.br
This prevents cluttering the output directory with a lot
of files from the extracted archive.
.PP
All extracted files are ensured that they are readable by the
current user.
.PP
If extracting an archive has an error in the middle of operation, patool does
not remove the temporary extraction directory
This directory whose name starts with \fBUnpack_\fP has all files that have been
extracted before the error.
.SS list
\fBpatool\fP \fBlist\fP [\fB\-\-password\fP \fIpassword\fP] <\fIarchive\fP>...
.TP
\fB\-\-password\fP \fIpassword\fP
Use given password for password-protected archives.
.PP
List files in archives.
.SS create
\fBpatool\fP \fBcreate\fP [\fB\-\-password\fP \fIpassword\fP] <\fIarchive\fP> <\fIfile-or-directory\fP>...
.TP
\fB\-\-password\fP \fIpassword\fP
Use given password for password-protected archives.
.PP
Create an archive from given files. All of the given files to add
to the archive must be readable by the current user.
The format of the archive to create is determined by the archive file
extension. If the archive program has options to maximize file compression,
\fBpatool\fP uses those options.
.SS test
\fBpatool\fP \fBtest\fP [\fB\-\-password\fP \fIpassword\fP] <\fIarchive\fP>...
.TP
\fB\-\-password\fP \fIpassword\fP
Use given password for password-protected archives.
.PP
Test the given archives. If the helper application does not support
testing, the archive contents are listed instead.
.SS diff
\fBpatool\fP \fBdiff\fP <\fIarchive1\fP> <\fIarchive2\fP>
.PP
Show differences between two archives with the \fBdiff(1)\fP program.
The diff options used are \fB\-urN\fP.
.SS search
\fBpatool\fP \fBsearch\fP [\fB\-\-password\fP \fIpassword\fP] <\fIpattern\fP> <\fIarchive\fP>
.TP
\fB\-\-password\fP \fIpassword\fP
Use given password for password-protected archives.
.PP
Unpack the given archive in a temporary directory and search in archive
contents for given pattern using the \fBgrep(1)\fP program.
The grep options used are \fB\-r\fP; additional options can be supplied
with the \fBGREP_OPTIONS\fP environment variable.
.SS repack
\fBpatool\fP \fBrepack\fP <\fIarchive\fP> <\fIarchive_new\fP>
.PP
Extract and re-compress archive to a different format.
The target archive format is determined by the file extension of \fIarchive_new\fP.
.SS formats
\fBpatool\fP \fBformats\fP
.PP
Show all supported archive formats (i.e. which helper applications
are available).
.SS version
\fBpatool\fP \fBversion\fP
.PP
Print version information.
.SH HELP OPTION
Specifying the help option \fB\-h\fP or \fB\-\-help\fP displays help for patool itself,
or a command.
.br
For example:
  \fBpatool \-h\fP - display help for patool
  \fBpatool extract \-\-help\fP - display help for the extract command
.SH AUTHOR
Bastian Kleineidam <bastian.kleineidam@web.de>
.SH COPYRIGHT
Copyright \(co 2010-2026 Bastian Kleineidam


debian/changelog
patool (1.8-0calvin1) unstable; urgency=low

  * New upstream release.

 -- Bastian Kleineidam <calvin@debian.org>  Sun, 19 Jul 2015 19:26:03 +0200

patool (1.7-1) unstable; urgency=low

  * New upstream release.

 -- Bastian Kleineidam <calvin@debian.org>  Fri, 27 Jun 2014 18:00:38 +0200

patool (1.6-1) unstable; urgency=low

  * New upstream release.

 -- Bastian Kleineidam <calvin@debian.org>  Sun, 08 Jun 2014 12:36:42 +0200

patool (1.5-1) unstable; urgency=low

  * New upstream release.

 -- Bastian Kleineidam <calvin@debian.org>  Sun, 30 Mar 2014 08:34:41 +0200

patool (1.4-2) unstable; urgency=low

  * Updated debian/watch

 -- Bastian Kleineidam <calvin@debian.org>  Wed, 15 Jan 2014 20:05:45 +0100

patool (1.4-1) unstable; urgency=low

  * New upstream release.

 -- Bastian Kleineidam <calvin@debian.org>  Fri, 26 Jul 2013 22:39:17 +0200

patool (1.3-1) unstable; urgency=low

  * New upstream release.

 -- Bastian Kleineidam <calvin@debian.org>  Fri, 19 Jul 2013 15:11:17 +0200

patool (1.2-1) unstable; urgency=low

  * New upstream release.
    (Closes: #711959)

 -- Bastian Kleineidam <calvin@debian.org>  Thu, 27 Jun 2013 22:20:13 +0200

patool (1.1-1) unstable; urgency=low

  * New upstream release.

 -- Bastian Kleineidam <calvin@debian.org>  Sun, 14 Apr 2013 14:41:58 +0200

patool (1.0-2) unstable; urgency=low

  * Start with a clean debian/changelog.
  * Remove unofficial remark from debian/control
  * Do not install custom bts file for reportbug.
  * Use machine-readable debian/copyright information.

 -- Bastian Kleineidam <calvin@debian.org>  Tue, 05 Mar 2013 19:44:45 +0100

patool (1.0-1) unstable; urgency=low

  * Official release (Closes: #702029)

 -- Bastian Kleineidam <calvin@debian.org>  Mon, 04 Mar 2013 19:12:32 +0100


patool/patool.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Copyright (C) 2010-2015 Bastian Kleineidam
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
"""
patool [global-options] {extract|list|create|diff|search|formats} [sub-command-options] <command-args>
"""
from __future__ import print_function
import sys
import argparse
import pydoc
import patoolib
from patoolib.util import log_error, log_internal_error, PatoolError
from patoolib.configuration import App

def run_extract(args):
    """Extract files from archive(s)."""
    res = 0
    for archive in args.archive:
        try:
            patoolib.extract_archive(archive, verbosity=args.verbosity, interactive=args.interactive, outdir=args.outdir, password=args.password)
        except PatoolError as msg:
            log_error("error extracting %s: %s" % (archive, msg))
            res += 1
    return res


def run_list(args):
    """List files in archive(s)."""
    res = 0
    for archive in args.archive:
        try:
            # increase default verbosity since the listing output should be visible
            verbosity = args.verbosity + 1
            patoolib.list_archive(archive, verbosity=verbosity, interactive=args.interactive, password=args.password)
        except PatoolError as msg:
            log_error("error listing %s: %s" % (archive, msg))
            res += 1
    return res


def run_test(args):
    """Test files in archive(s)."""
    res = 0
    for archive in args.archive:
        try:
            patoolib.test_archive(archive, verbosity=args.verbosity, interactive=args.interactive, password=args.password)
        except PatoolError as msg:
            log_error("error testing %s: %s" % (archive, msg))
            res += 1
    return res


def run_create(args):
    """Create an archive from given files."""
    res = 0
    try:
        patoolib.create_archive(args.archive, args.filename, verbosity=args.verbosity, interactive=args.interactive, password=args.password)
    except PatoolError as msg:
        log_error("error creating %s: %s" % (args.archive, msg))
        res = 1
    return res


def run_diff(args):
    """Show differences between two archives."""
    try:
        res = patoolib.diff_archives(args.archive1, args.archive2, verbosity=args.verbosity, interactive=args.interactive)
    except PatoolError as msg:
        log_error("error showing differences between %s and %s: %s" % (args.archive1, args.archive2, msg))
        res = 2
    return res


def run_search(args):
    """Search for pattern in given archive."""
    try:
        res = patoolib.search_archive(args.pattern, args.archive, verbosity=args.verbosity, interactive=args.interactive, password=args.password)
    except PatoolError as msg:
        log_error("error searching %s: %s" % (args.archive, msg))
        res = 2
    return res


def run_repack(args):
    """Repackage one archive in another format."""
    res = 0
    try:
        patoolib.repack_archive(args.archive_src, args.archive_dst, verbosity=args.verbosity, interactive=args.interactive)
    except PatoolError as msg:
        log_error("error repacking %s: %s" % (args.archive_src, msg))
        res = 1
    return res


def run_recompress(args):
    """Recompress an archive to smaller size."""
    res = 0
    try:
        patoolib.recompress_archive(args.archive, verbosity=args.verbosity, interactive=args.interactive, password=args.password)
    except PatoolError as msg:
        log_error("error recompressing %s: %s" % (args.archive, msg))
        res = 1
    return res


def run_formats (args):
    """List supported and available archive formats."""
    patoolib.list_formats()
    return 0


class ArgumentParser(argparse.ArgumentParser):
    """Custom argument parser."""

    def print_help(self, file=None):
        """Paginate help message on TTYs."""
        msg = self.format_help()
        if file is None:
            file = sys.stdout
        if hasattr(file, "isatty") and file.isatty():
            pydoc.pager(msg)
        else:
            print(msg, file=file)

Examples = """\
EXAMPLES
  patool extract archive.zip otherarchive.rar
  patool --verbose test dist.tar.gz
  patool list package.deb
  patool --verbose create myfiles.zip file1.txt dir/
  patool diff release1.0.tar.xz release2.0.zip
  patool search "def urlopen" python-3.3.tar.gz
  patool repack linux-2.6.33.tar.gz linux-2.6.33.tar.bz2
  patool recompress images.zip
"""

Version = """\
VERSION
  %s
""" % App

def create_argparser():
    """Construct and return an argument parser."""
    epilog = Examples + "\n" + Version
    parser = ArgumentParser(description="An archive file manager.",
        epilog=epilog, formatter_class=argparse.RawDescriptionHelpFormatter)
    parser.add_argument('--verbose', '-v', action='count', default=0, dest='verbosity', help="verbose operation; can be given multiple times")
    parser.add_argument('--non-interactive', dest='interactive', default=True, action='store_false',
        help="don't query for user input (ie. passwords or when overwriting duplicate files); use with care since overwriting files or ignoring passwords could be unintended")
    subparsers = parser.add_subparsers(help='the archive command; type "patool COMMAND -h" for command-specific help', dest='command')
    # extract
    parser_extract = subparsers.add_parser('extract', help='extract one or more archives')
    parser_extract.add_argument('--outdir', help="output directory to extract to")
    parser_extract.add_argument('--password', help="password for encrypted files")
    parser_extract.add_argument('archive', nargs='+', help="an archive file")
    # list
    parser_list = subparsers.add_parser('list', help='list members or one or more archives')
    parser_list.add_argument('--password', help="password for encrypted files")
    parser_list.add_argument('archive', nargs='+', help="an archive file")
    # create
    parser_create = subparsers.add_parser('create', help='create an archive')
    parser_create.add_argument('--password', help="password to encrypt files")
    parser_create.add_argument('archive', help="the archive file; the file extension determines the archive program")
    parser_create.add_argument('filename', nargs='+', help="a file or directory to add to the archive; note that some archive programs do not support directories")
    # test
    parser_test = subparsers.add_parser('test', help='test an archive')
    parser_test.add_argument('--password', help="password for encrypted files")
    parser_test.add_argument('archive', nargs='+', help='an archive file')
    # repack
    parser_repack = subparsers.add_parser('repack', help='repack an archive to a different format')
    parser_repack.add_argument('archive_src', help='source archive file')
    parser_repack.add_argument('archive_dst', help='target archive file')
    # recompress
    parser_recompress = subparsers.add_parser('recompress', help='recompress an archive to smaller size')
    parser_recompress.add_argument('--password', help="password for encrypted files")
    parser_recompress.add_argument('archive', help='an archive file')
    # diff
    parser_diff = subparsers.add_parser('diff', help='show differences between two archives')
    parser_diff.add_argument('archive1', help='the first archive file')
    parser_diff.add_argument('archive2', help='the second archive file')
    # search
    parser_search = subparsers.add_parser('search', help="search contents of archive members")
    parser_search.add_argument('--password', help="password for encrypted files")
    parser_search.add_argument('pattern', help='the grep(1) compatible search pattern')
    parser_search.add_argument('archive', help='the archive file')
    # formats
    subparsers.add_parser('formats', help="show supported archive formats")
    # optional bash completion
    try:
        import argcomplete
        argcomplete.autocomplete(parser)
    except ImportError:
        pass
    return parser


def main():
    """Parse options and execute commands."""
    try:
        argparser = create_argparser()
        args = argparser.parse_args()
        if args.command is None:
            # Python 3.3.1 under linux allows an empty command somehow
            argparser.error("too few arguments")
        # run subcommand function
        res = globals()["run_%s" % args.command](args)
    except KeyboardInterrupt:
        log_error("aborted")
        res = 1
    except Exception:
        log_internal_error()
        res = 2
    return res


if __name__ == '__main__':
    sys.exit(main())
